File: japan-geoid-py/examples/convert_asc_to_bin.py
"""
$ python3 convert_to_bin.py ./gsigeo2011_ver2_2.asc
"""

import gzip
import sys
from pathlib import Path

from japan_geoid import GsiGeoid

if __name__ == "__main__":
    path = Path(sys.argv[1])

    # オリジナルのASCII形式のジオイドモデルを読み込む
    with open(path, "r") as f:
        geoid = GsiGeoid.from_ascii(f.read())

    # ジオイド高を取得する
    (lng, lat) = (138.2839817085188, 37.12378643088312)
    height = geoid.get_height(lng, lat)
    print(f"{lng=} {lat=} {height=}")

    # ジオイドモデルをバイナリ形式で保存する
    with gzip.open(path.with_suffix(".bin.gz"), "wb") as dest_f:
        dest_f.write(geoid.to_binary())

    # バイナリ形式のジオイドモデルを読み込む
    with gzip.open(path.with_suffix(".bin.gz"), "rb") as f:
        geoid = GsiGeoid.from_binary(f.read())

    # ジオイド高を取得する
    (lng, lat) = (138.2839817085188, 37.12378643088312)
    height = geoid.get_height(lng, lat)
    print(f"{lng=} {lat=} {height=}")

File: examples/convert_asc_to_bin.rs
use flate2::{read::GzDecoder, write::GzEncoder};
use std::{
    fs::File,
    io::{BufReader, BufWriter},
};

use japan_geoid::{Geoid, MemoryGrid};

fn main() -> std::io::Result<()> {
    let (lng, lat) = (138.2839817085188, 37.12378643088312);

    // Load from the original ascii format.
    let file = File::open("./gsigeo2011_ver2_2.asc")?;
    let mut reader = BufReader::new(file);
    let geoid = MemoryGrid::from_ascii_reader(&mut reader)?;

    let z = geoid.get_height(lng, lat);
    println!("Input: (lng: {}, lat: {}) -> Geoid height: {}", lng, lat, z);

    // Dump as the binary format.
    let file = File::create("./gsigeo2011_ver2_2.bin.gz")?;
    let mut writer = GzEncoder::new(BufWriter::new(file), flate2::Compression::fast());
    geoid.to_binary_writer(&mut writer)?;
    writer.finish()?;

    // Load from the binary.
    let mut file = File::open("./gsigeo2011_ver2_2.bin.gz")?;
    let mut reader = GzDecoder::new(&mut file);
    let geoid = MemoryGrid::from_binary_reader(&mut reader)?;

    let z = geoid.get_height(lng, lat);
    println!("Input: (lng: {}, lat: {}) -> Geoid height: {}", lng, lat, z);

    Ok(())
}
